### PR TITLE
Add multi-build support in a single module

### LINKED ISSUE #16053: Add multi-build support in a single module
## Description

Give the possibility to build the same module multiple times with different profiles, generating different artifacts.

## Implementation ideas

Add multiple executions to the Quarkus maven plugin, specifying a different Quarkus profile for each execution.
The plugin could then generate one artifact for each execution and store it into a specific output directory.

Here is an example of the `executions` we could find in the `pom.xml`, resulting in 2 subfolders:
- `target/foo-quarkus-app`
- `target/bar-quarkus-app`

```
<executions>
  <execution>
    <id>build-foo</id>
    <goals>
      <goal>generate-code</goal>
      <goal>generate-code-tests</goal>
      <goal>build</goal>
    </goals>
    <configuration>
      <systemProperties>
        <quarkus.profile>foo</quarkus.profile>
        <quarkus.package.output-directory>foo-quarkus-app</quarkus.package.output-directory>
      </systemProperties>
    </configuration>
  </execution>
  <execution>
    <id>build-bar</id>
    <goals>
      <goal>generate-code</goal>
      <goal>generate-code-tests</goal>
      <goal>build</goal>
    </goals>
    <configuration>
      <systemProperties>
        <quarkus.profile>bar</quarkus.profile>
        <quarkus.package.output-directory>bar-quarkus-app</quarkus.package.output-directory>
      </systemProperties>
    </configuration>
  </execution>
</executions>
```

### PR DESCRIPTION
**Pull request by bcournaud**
fixes #16053

## Motivations

Be able to build the same module multiple times with different profiles, generating different artifacts.

## Modifications

A new parameter has been added to the `BuildMojo` that contains the system properties of the plugin.
These properties are added to the global system properties if they are not already defined.

Once the build is done, these plugin properties are removed from the system properties so that they are not taken into account by another build which may have different properties.

All the plugin system properties are now used during the execution 

_Note that if multiple executions are defined, the `quarkus.package.output-directory` property must be defined and different for each execution in order to store the resulting artifact in a specific folder for each execution (instead of overriding the artifacts at each build)._

### CODE REVIEW

**Comment by aloubyansky** [comment]
I think it'd be better instead of setting system properties to turn them into the build properties, i.e. add the properties parameter to https://github.com/quarkusio/quarkus/blob/main/devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapMojo.java and set them like https://github.com/quarkusio/quarkus/blob/main/devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java#L139

WDYT?

Thanks for the great contribution, btw!

Just in case, fyi, @famod

**Comment by famod** [comment]
Yeah, great contribution!
I'm wondering if that could be used to build for multiple databases, given that you cannot differentiate dependencies with this approach (but you could with profiles).

**Comment by essobedo** [comment]
@aloubyansky 
> I think it'd be better instead of setting system properties to turn them into the build properties, i.e. add the properties parameter to https://github.com/quarkusio/quarkus/blob/main/devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapMojo.java and set them like https://github.com/quarkusio/quarkus/blob/main/devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java#L139
> 
> WDYT?
I see, it makes sense but do we agree that those properties will be set as system properties at the end? Indeed, otherwise a parameter like the Quarkus profile won't be taken into account and setting the Quarkus profile is necessary to be able to fulfill the requirements of this feature request.

@famod
> I'm wondering if that could be used to build for multiple databases, given that you cannot differentiate dependencies with this approach (but you could with profiles).
Good point, I'm afraid that indeed this approach won't help much in your case since the dependencies will be the same, as you mentioned (maybe a little in native mode? if it removes dead code, not sure and not tested) however it could be interesting as the next step of this improvement, WDYT?

**Comment by aloubyansky** [comment]
> @aloubyansky
> 
> > I think it'd be better instead of setting system properties to turn them into the build properties, i.e. add the properties parameter to https://github.com/quarkusio/quarkus/blob/main/devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapMojo.java and set them like https://github.com/quarkusio/quarkus/blob/main/devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java#L139
> > WDYT?
> 
> I see, it makes sense but do we agree that those properties will be set as system properties at the end? Indeed, otherwise a parameter like the Quarkus profile won't be taken into account and setting the Quarkus profile is necessary to be able to fulfill the requirements of this feature request.

Besides the Quarkus profile one, would the rest work as project properties for you? Or perhaps in that case you could model all that with Maven profiles instead?

I haven't checked, would setting system properties break in case of parallel builds?

**Comment by essobedo** [comment]
@aloubyansky
> Besides the Quarkus profile one, would the rest work as project properties for you? Or perhaps in that case you could model all that with Maven profiles instead?

We will test it asap with @bcournaud 

> 
> I haven't checked, would setting system properties break in case of parallel builds?

Good question, not tested either but I guess it should not work, however AFAIK custom Quarkus profile can only be set thanks to a System property, a environment variable or a static field, in other words there isn't any easy way to isolate the value unless we use a dedicated ClassLoader per build, or we make `ProfileManager` evolve to allow to set the profile into a `ThreadLocal` or we simply consider it as the current limitation?

**Comment by aloubyansky** [comment]
Right, which is why I am hesitating to set system properties by default. Quarkus profile is a known exception for now.

File: devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java [hunk 1]
```diff
@@ -262,13 +265,11 @@ private AppArtifact initAppArtifact(QuarkusBootstrapMojo mojo) throws MojoExecut
...
         @Override
-        public void close() throws IOException {
-            if (curatedApp != null) {
-                curatedApp.close();
-                curatedApp = null;
-            }
-            this.artifactResolver = null;
-            this.quarkusBootstrap = null;
+        public void close() {
+            // TODO It seems this close method is never call.
```
Comment (aloubyansky) [inline devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java:269]: It should be. Do you have `<extensions>true</extensions>` in the `quarkus-maven-plugin` config?

Comment (bcournaud) [reply devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java:269]: Good catch! I was not aware of this configuration. It might make the modifications in `QuarkusBootstrapProvider` useless (I will check).

**Comment by bcournaud** [comment]
Thanks for the suggestions @aloubyansky.
We kept only the **Quarkus Profile** in the `<systemProperties>` and moved the rest (`quarkus.package.output-directory`) into a dedicated `<properties>` section which will not be added to the system properties.

File: devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java (hunk 1 above)
Comment (bcournaud) [reply devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java:269]: So the `close()` is indeed properly called. I will remove the comment.
We still have to do the change in the `QuarkusBootstrapProvider`[1] though, as the `close()` is called only once at the end of the maven execution.

[1] (the change)[https://github.com/quarkusio/quarkus/pull/16054/files#diff-a3d8d753534cdccadf2cf2437e8d090dd79e67832603e99486696369e8328f62R39] allows to store multiple instances of `QuarkusAppBootstrapProvider`, each of them containing specific properties defined in each `<execution>` of the plugin.

**Comment by aloubyansky** [comment]
@bcournaud Thanks! Could you please squash the commits and un-draft it please?

File: devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java [hunk 2]
```diff
@@ -30,7 +34,7 @@
 /**
  * Builds the Quarkus application.
  */
-@Mojo(name = "build", defaultPhase = LifecyclePhase.PACKAGE, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME, threadSafe = true)
```
Comment (essobedo) [inline devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: We mark it as non thread safe as long as the problem with the quarkus profile (set as a System property) is not fixed

Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: Just curious, have you actually verified the system property set in one Mojo is visible in another one? You haven't I will check that.

Comment (famod) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: You can certainly re-check this but in the past I have filed multiple issues for various maven plugins that fell in the same trap, e.g. NPEs during iterating the properties while another Maven thread added or removed a property.

Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: Was that project properties of system properties? Thanks for the info @famod

Comment (famod) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: System properties, AFAIR.

Comment (essobedo) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: @aloubyansky yes I confirm that I have tested with a project with two modules, in each module I have 2 executions, in the first module, the first execution sets the Quarkus profile to `foo` and the second one sets the Quarkus profile to `bar` and in the second module, it does the opposite (first `bar` and second `foo`). I launched ` mvn -T 1C clean install` and ended up with a build success but incorrect Quarkus profile set at runtime (Quarkus profile set to `bar` instead of `foo` and vice versa). 

In other words, I confirm that system properties make the plugin non thread safe unless you use the same system properties across all your modules.

Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: Thanks a lot @essobedo

Comment (essobedo) [reply devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:33]: @aloubyansky If you want to keep it thread safe, we can try to investigate more to see how we can isolate the Quarkus profile (by relying on the ClassLoader for example) but it won't be easy

**Comment by aloubyansky** [comment]
The CI failures don't seem to be related.

**Review by aloubyansky — APPROVED** [review]

**Comment by aloubyansky** [comment]
Thank you @bcournaud and @essobedo for another great contribution!

**Comment by famod** [comment]
Is there any follow up issue tracking the `threadSafe` topic? Building Quarkus in parallel really spams the console with that thread safety warning...

**Comment by essobedo** [comment]
@famod We did not create any ticket for this topic. If you create one, please let me know, I can have a look asap.
